File: pdftract-dotnet/tests/Pdftract.Tests/DiagnosticSerializationTests.cs
using System.Text.Json;
using Pdftract.Models;
using Xunit;

namespace Pdftract.Tests;

public class DiagnosticSerializationTests
{
    [Fact]
    public void Diagnostic_PreservesCanonicalFieldsAndOmitsAbsentContext()
    {
        const string json = """
            {"code":"STREAM_DECODE_ERROR","message":"zlib stream truncated mid-inflation","severity":"warning","page_index":3,"location":{"object_number":42,"generation_number":7},"hint":"Inspect the source PDF for corrupt stream data"}
            """;

        var diagnostic = JsonSerializer.Deserialize(
            json, PdftractJsonContext.Default.Error);

        Assert.NotNull(diagnostic);
        Assert.Equal("STREAM_DECODE_ERROR", diagnostic.Code);
        Assert.Equal("warning", diagnostic.Severity);
        Assert.Equal(3, diagnostic.PageIndex);
        Assert.Equal(42, diagnostic.Location!.ObjectNumber);
        Assert.Equal(7, diagnostic.Location.GenerationNumber);
        Assert.Equal("Inspect the source PDF for corrupt stream data", diagnostic.Hint);

        var encoded = JsonSerializer.Serialize(
            diagnostic, PdftractJsonContext.Default.Error);
        using var document = JsonDocument.Parse(encoded);
        Assert.Equal(3, document.RootElement.GetProperty("page_index").GetInt32());
        Assert.Equal(42, document.RootElement.GetProperty("location")
            .GetProperty("object_number").GetInt32());

        var bare = new Error
        {
            Code = "XREF_REPAIRED",
            Message = "repaired",
            Severity = "info"
        };
        using var bareDocument = JsonDocument.Parse(JsonSerializer.Serialize(
            bare, PdftractJsonContext.Default.Error));
        Assert.False(bareDocument.RootElement.TryGetProperty("page_index", out _));
        Assert.False(bareDocument.RootElement.TryGetProperty("location", out _));
        Assert.False(bareDocument.RootElement.TryGetProperty("hint", out _));
    }
}

File: pdftract-swift/Tests/PdftractTests/DiagnosticSerializationTests.swift
import XCTest
@testable import Pdftract

final class DiagnosticSerializationTests: XCTestCase {
    func testDiagnosticPreservesCanonicalFieldsAndOmitsAbsentContext() throws {
        let input = """
        {"code":"STREAM_DECODE_ERROR","message":"zlib stream truncated mid-inflation","severity":"warning","page_index":3,"location":{"object_number":42,"generation_number":7},"hint":"Inspect the source PDF for corrupt stream data"}
        """.data(using: .utf8)!

        let diagnostic = try JSONDecoder().decode(Diagnostic.self, from: input)
        XCTAssertEqual(diagnostic.code, "STREAM_DECODE_ERROR")
        XCTAssertEqual(diagnostic.severity, "warning")
        XCTAssertEqual(diagnostic.pageIndex, 3)
        XCTAssertEqual(diagnostic.location?.objectNumber, 42)
        XCTAssertEqual(diagnostic.location?.generationNumber, 7)
        XCTAssertEqual(diagnostic.hint, "Inspect the source PDF for corrupt stream data")

        let encoded = try JSONSerialization.jsonObject(
            with: JSONEncoder().encode(diagnostic)) as! [String: Any]
        XCTAssertEqual(encoded["page_index"] as? Int, 3)
        XCTAssertEqual((encoded["location"] as? [String: Any])?["object_number"] as? Int, 42)

        let bare = Diagnostic(code: "XREF_REPAIRED", message: "repaired", severity: "info")
        let bareObject = try JSONSerialization.jsonObject(
            with: JSONEncoder().encode(bare)) as! [String: Any]
        XCTAssertNil(bareObject["page_index"])
        XCTAssertNil(bareObject["location"])
        XCTAssertNil(bareObject["hint"])
    }
}
